Panic occurs while opening the corrupt boltDB which leads to unrecoverable etcd pod crashed.

**Describe the bug:**
It has been observed that when `db file` of etcd data-dir has been corrupted and when backup-restore tries to [verify db](https://github.com/gardener/etcd-backup-restore/blob/f6e1b93e58cf6e6c1d341a22e7e47c9e6cfd7d8a/pkg/initializer/validator/datavalidator.go#L268) by [opening the boltDB](https://github.com/gardener/etcd-backup-restore/blob/f6e1b93e58cf6e6c1d341a22e7e47c9e6cfd7d8a/pkg/initializer/validator/datavalidator.go#L403) then [panic occurs in BoltDB](https://github.com/etcd-io/bbolt/blob/master/cursor.go#L250) as BoltDB is not expecting a corrupted DB file as it contains the information of pages and other metadata information.
All this leads to etcd and backup-restore into an unrecoverable state.

**Expected behavior:**
If `db file` of etcd data-dir has been corrupted then backup-restore should detect this and should be able to restore the data-dir of etcd in case of cluster size=1 and in case of cluster size>1 then backup-restore should able to trigger the single member restoration scenario.


**How To Reproduce (as minimally and precisely as possible):**
1. Starts an Etcd.
2. Put some dummy data into the etcd.
3. In order to corrupt the db file: open a DB file of Etcd in any editor and write some dummy data.
4. you might have to try few times to reproduce this error.

**Logs:**
Backup-restore logs:
```
INFO[0223] Checking for data directory structure validity...
INFO[0223] Get the latest Etcd revision...
panic: invalid page type: 0: 0

goroutine 1681 [running]:
go.etcd.io/bbolt.(*Cursor).search(0xc0008838e8, {0xc000883945, 0x3, 0x3}, 0xc000883880)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/vendor/go.etcd.io/bbolt/cursor.go:250 +0x28c
go.etcd.io/bbolt.(*Cursor).seek(0xc0008838e8, {0xc000883945, 0x28e23c0, 0xc000780001})
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/vendor/go.etcd.io/bbolt/cursor.go:159 +0x48
go.etcd.io/bbolt.(*Bucket).Bucket(0xc000c00018, {0xc000883945, 0x0, 0x3})
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/vendor/go.etcd.io/bbolt/bucket.go:105 +0xc8
go.etcd.io/bbolt.(*Tx).Bucket(...)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/vendor/go.etcd.io/bbolt/tx.go:101
github.com/gardener/etcd-backup-restore/pkg/initializer/validator.getLatestEtcdRevision.func1(0xc000385b20)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/initializer/validator/datavalidator.go:413 +0x45
go.etcd.io/bbolt.(*DB).View(0xc000385b20, 0xc000883a60)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/vendor/go.etcd.io/bbolt/db.go:725 +0x82
github.com/gardener/etcd-backup-restore/pkg/initializer/validator.getLatestEtcdRevision({0xc000385b20, 0x1b})
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/initializer/validator/datavalidator.go:412 +0x190
github.com/gardener/etcd-backup-restore/pkg/initializer/validator.(*DataValidator).sanityCheck(0xc00068db00, 0xffffffffffffffff)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/initializer/validator/datavalidator.go:145 +0x3f7
github.com/gardener/etcd-backup-restore/pkg/initializer/validator.(*DataValidator).Validate(0xc00068db00, {0x293a893, 0x4}, 0x2629080)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/initializer/validator/datavalidator.go:69 +0x46
github.com/gardener/etcd-backup-restore/pkg/initializer.(*EtcdInitializer).Initialize(0xc00010fde8, {0x293a893, 0x4}, 0x4)
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/initializer/initializer.go:68 +0x125
github.com/gardener/etcd-backup-restore/pkg/server.(*HTTPHandler).serveInitialize.func1()
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/server/httpAPI.go:266 +0x511
created by github.com/gardener/etcd-backup-restore/pkg/server.(*HTTPHandler).serveInitialize
	/Users/i539698/go/src/backup_forked/etcd-backup-restore/pkg/server/httpAPI.go:230 +0x245
```



**Anything else we need to know?**:
This error occurs due to BoltDB says it shouldn't get corrupted.

--- a/pkg/initializer/validator/datavalidator.go
+++ b/pkg/initializer/validator/datavalidator.go
@@ -54,6 +54,9 @@ var (
 
 	// ErrCorrupt is returned when a checking a data file finds errors.
 	ErrCorrupt = errors.New("invalid value")
+
+	// isBoltDBPanic is a flag which indicates the whether the panic occurs while opening the Bolt database.
+	isBoltDBPanic = false
 )
 
 func (d *DataValidator) memberDir() string { return filepath.Join(d.Config.DataDir, "member") }
@@ -141,6 +144,7 @@ func (d *DataValidator) sanityCheck(failBelowRevision int64) (DataDirStatus, err
 		return DataDirectoryValid, nil
 	}
 
+	d.Logger.Info("Checking for Etcd revision...")
 	etcdRevision, err := getLatestEtcdRevision(d.backendPath())
 	if err != nil && errors.Is(err, bolt.ErrTimeout) {
 		d.Logger.Errorf("another etcd process is using %v and holds the file lock", d.backendPath())
@@ -148,6 +152,13 @@ func (d *DataValidator) sanityCheck(failBelowRevision int64) (DataDirStatus, err
 	} else if err != nil {
 		d.Logger.Infof("unable to get current etcd revision from backend db file: %v", err)
 		return DataDirectoryCorrupt, nil
+	}
+
+	if isBoltDBPanic {
+		d.Logger.Info("Bolt database panic: database file found to be invalid.")
+		// reset the isBoltDBPanic
+		isBoltDBPanic = false
+		return BoltDBCorrupt, nil
 	}
 
 	if d.OriginalClusterSize > 1 {
@@ -205,6 +216,12 @@ func (d *DataValidator) checkForDataCorruption() error {
 			return err
 		}
 		return fmt.Errorf("invalid db files: %v", err)
+	}
+	if isBoltDBPanic {
+		d.Logger.Info("Bolt database panic: database file found to be invalid.")
+		// reset the isBoltDBPanic
+		isBoltDBPanic = false
+		return fmt.Errorf("invalid db files")
 	}
 	return nil
 }
@@ -272,6 +289,13 @@ func verifyDB(path string) error {
 	} else if _, err := os.Stat(path); os.IsNotExist(err) {
 		return ErrFileNotFound
 	}
+
+	defer func() {
+		if err := recover(); err != nil {
+			// set the flag: isBoltDBPanic
+			isBoltDBPanic = true
+		}
+	}()
 
 	// Open database.
 	db, err := bolt.Open(path, 0666, &bolt.Options{Timeout: timeoutToOpenBoltDB})
@@ -376,7 +400,7 @@ waitLoop:
 		case <-timer.C:
 			break waitLoop
 		default:
-			latestSyncedEtcdRevision, _ = getLatestSyncedRevision(clientKV)
+			latestSyncedEtcdRevision, _ = getLatestSyncedRevision(clientKV, d.Logger)
 			if latestSyncedEtcdRevision >= latestSnapshotRevision {
 				d.Logger.Infof("After starting embeddedEtcd backend DB file revision (%d) is greater than or equal to latest snapshot revision (%d): no data loss", latestSyncedEtcdRevision, latestSnapshotRevision)
 				break waitLoop
@@ -399,6 +423,13 @@ func getLatestEtcdRevision(path string) (int64, error) {
 	if _, err := os.Stat(path); err != nil {
 		return -1, fmt.Errorf("unable to stat backend db file: %v", err)
 	}
+
+	defer func() {
+		if err := recover(); err != nil {
+			// set the flag: isBoltDBPanic
+			isBoltDBPanic = true
+		}
+	}()
 
 	db, err := bolt.Open(path, 0400, &bolt.Options{Timeout: timeoutToOpenBoltDB, ReadOnly: true})
 	if err != nil {
@@ -433,14 +464,14 @@ func getLatestEtcdRevision(path string) (int64, error) {
 }
 
 // getLatestSyncedRevision finds out the latest revision on etcd db file when embedded etcd is started to double check the latest revision of etcd db file.
-func getLatestSyncedRevision(client client.KVCloser) (int64, error) {
+func getLatestSyncedRevision(client client.KVCloser, logger *logrus.Logger) (int64, error) {
 	var latestSyncedRevision int64
 
 	ctx, cancel := context.WithTimeout(context.TODO(), connectionTimeout)
 	defer cancel()
 	resp, err := client.Get(ctx, "", clientv3.WithLastRev()...)
 	if err != nil {
-		fmt.Printf("Failed to get the latest etcd revision: %v\n", err)
+		logger.Errorf("failed to get the latest etcd revision: %v\n", err)
 		return latestSyncedRevision, err
 	}
 	latestSyncedRevision = resp.Header.Revision

--- a/pkg/initializer/validator/types.go
+++ b/pkg/initializer/validator/types.go
@@ -36,6 +36,8 @@ const (
 	DataDirectoryInvStruct
 	// DataDirectoryCorrupt indicates data directory is corrupt.
 	DataDirectoryCorrupt
+	// BoltDBCorrupt indicates Bolt database is corrupt.
+	BoltDBCorrupt
 	// DataDirectoryStatusUnknown indicates validator failed to check the data directory status.
 	DataDirectoryStatusUnknown
 	// DataDirStatusInvalidInMultiNode indicates validator failed to check the data directory status in multi-node etcd cluster.

--- a/pkg/initializer/validator/validator_suite_test.go
+++ b/pkg/initializer/validator/validator_suite_test.go
@@ -38,12 +38,6 @@ var (
 	etcdRevision int64
 )
 
-// fileInfo holds file information such as file name and file path
-type fileInfo struct {
-	name string
-	path string
-}
-
 func TestValidator(t *testing.T) {
 	RegisterFailHandler(Fail)
 	RunSpecs(t, "Validator Suite")

